mycpp/leaky_types.cc

```diff
@@ -2,16 +2,144 @@
   #include "mycpp/mylib_old.h"
 using gc_heap::StackRoots;  // no-op
 using mylib::AllocStr;
+using mylib::CopyBufferIntoNewStr;
 #else
   #include "mycpp/gc_builtins.h"
   #include "mycpp/gc_types.h"
 using gc_heap::kEmptyString;
 using gc_heap::StackRoots;
 using gc_heap::Str;
+using gc_heap::CopyBufferIntoNewStr;
 #endif
 
 #include <ctype.h>  // isalpha(), isdigit()
 
+// Translation of Python's print().
+void print(Str* s) {
+  fputs(s->data(), stdout);
+  fputs("\n", stdout);
+}
+
+// Like print(..., file=sys.stderr), but Python code explicitly calls it.
+void println_stderr(Str* s) {
+  fputs(s->data(), stderr);
+  fputs("\n", stderr);
+}
+
+Str* str_repeat(Str* s, int times) {
+  // Python allows -1 too, and Oil used that
+  if (times <= 0) {
+    return kEmptyString;
+  }
+  int len_ = len(s);
+  int new_len = len_ * times;
+  char* data = static_cast<char*>(malloc(new_len + 1));
+
+  char* dest = data;
+  for (int i = 0; i < times; i++) {
+    memcpy(dest, s->data_, len_);
+    dest += len_;
+  }
+  data[new_len] = '\0';
+  return CopyBufferIntoNewStr(data, new_len);
+}
+
+// Helper for str_to_int() that doesn't use exceptions.
+// Like atoi(), but with better error checking.
+bool _str_to_int(Str* s, int* result, int base) {
+  int s_len = len(s);
+  if (s_len == 0) {
+    return false;  // special case for empty string
+  }
+
+  char* p;  // mutated by strtol
+
+  long v = strtol(s->data(), &p, base);
+  switch (v) {
+  case LONG_MIN:
+    // log("underflow");
+    return false;
+  case LONG_MAX:
+    // log("overflow");
+    return false;
+  }
+
+  *result = v;
+
+  // Return true if it consumed ALL characters.
+  const char* end = s->data_ + s_len;
+
+  // log("start %p   p %p   end %p", s->data_, p, end);
+  if (p == end) {
+    return true;
+  }
+
+  // Trailing space is OK!
+  while (p < end) {
+    if (!isspace(*p)) {
+      return false;
+    }
+    p++;
+  }
+  return true;
+}
+
+// for os_path.join()
+// NOTE(Jesse): Perfect candidate for bounded_buffer
```

Comment:
Yes agree, we're seeing that pattern a lot

I think it makes sense to have a debug mode bounds checker, and then have it of in opt mode